### ISSUE TITLE
Map-style DataPipe to read from s3

### ISSUE THREAD

**Issue by gombru** [pytorch/data#707]
### 🚀 The feature

[Amazon S3 plugin for PyTorch ](https://aws.amazon.com/blogs/machine-learning/announcing-the-amazon-s3-plugin-for-pytorch/)proposes S3Dataset which is a Map-style PyTorch Dataset. I was looking for a similar feature in torchdata but only found [S3FileLoader](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.S3FileLoader.html#torchdata.datapipes.iter.S3FileLoader) which doesn't meet my requirements.
 
 Is there any implementation of a Map-style DataPipe I am missing? Or any method to do a similar thing with the existing tools?
 
The main requirement is that I need to read images from s3, apply a transformation, and keep them syncronized with a list of labels. 

 Thank you

### Motivation, pitch

Map-style DataPipe to read from s3 to complement the existing itetable style Datapipe to read from s3.

**Comment by NivekT** [comment]
It is possible for you to use the Iter-style S3 DataPipes, perform your transformation, then convert it to Map-style later with [`.to_map_datapipe()`](https://pytorch.org/data/beta/generated/torchdata.datapipes.map.IterToMapConverter.html#torchdata.datapipes.map.IterToMapConverter)?

**Comment by gombru** [comment]
Maybe that would work. Does [.to_map_datapipe()](https://pytorch.org/data/beta/generated/torchdata.datapipes.map.IterToMapConverter.html#torchdata.datapipes.map.IterToMapConverter) allow to keep an idx as Map-style DataPipes, so that then I can fetch the corresponding label from the list?

Do you have any example of that?
Thank you

**Comment by NivekT** [comment]
Something like this works:

```python
# 1 - `dp` has sequence of length 2, first element will become the index
dp = IterableWrapper([('a', 1), ('b', 2), ('c', 1)])
mdp = dp.to_map_datapipe()
print(mdp['a'])  # 1

# 2 - `dp` doesn't have seq of length 2, need to pass in an extra function
def row_to_tuple(row):
    label = row[0]
    data = row[1:]
    return label, data

dp2 = IterableWrapper([('a', 1, 1, 1, 1, 1, 1), ('b', 2, 2, 2, 2, 2, 2), ('c', 3, 3, 3, 3, 3, 3)])
mdp2 = dp2.to_map_datapipe(key_value_fn=row_to_tuple)
print(mdp2['a'])  # (1, 1, 1, 1, 1, 1)
```

**Comment by gombru** [comment]
Thanks for the help.

I found that the best solution for my setup here was store the labels as a dict where keys are (an ID in) the images url, so that then I can easily find them in __iter__.

A question related to [S3FileLoader](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.S3FileLoader.html#torchdata.datapipes.iter.S3FileLoader): Is there a way to ignore missing files?
S3FileLoader reads data in the background and breaks when a file is missing. In a Map-style dataloader I could handle missing files in the loader; is there a way to do it here?

I could check for missing files previously, but some of them might be deleted during training.

**Comment by NivekT** [comment]
> A question related to [S3FileLoader](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.S3FileLoader.html#torchdata.datapipes.iter.S3FileLoader): Is there a way to ignore missing files?

I think we would prefer to filter those files out ahead of time with something like `filter` to check if the file exists before trying to load it. We generally don't expect that to happen since a common pattern is to list out the files on S3 first (with `.list_files_by_s3()`),

I'm unsure about the case where files may be deleted during training. What is causing that to happen? Is that intended?
One concern I have is - if that happens, then you will have different numbers of files/samples in different epochs of your training, and that may be problematic?

cc: @ejguan

**Comment by gombru** [comment]
Yes I'll filter the files out before training, but still a file can be deleted during training. That's expected and not related to the training itself but to how the database is maneged (can't replicate it because of the size).  

In a Map-Style loader I could check if the file existed in each __getitem__ but with this approach I cannot. So wanted to check if there's any alternative.

**Comment by NivekT** [comment]
I believe it is possible for you to check if file exists within `__iter__` of an IterDataPipe.

@ejguan can you confirm? Any thoughts?

**Comment by ejguan** [comment]
Like @NivekT said, you might want to implement a special `DataPipe`. And, within `__iter__`, you should suppress Error for those missing files.
```py
class SkipMissingFile(IterDataPipe):
    def __init__(self, dp):
        self.dp = dp
    def __iter__(self):
        it = iter(self.dp)
        while True:
            try:
                d = next(it)
            except MissingFileException:
                 pass
            except StopIteration:
                 break
            yield d
```
Does this work for you?

**Comment by gombru** [comment]
I solved it in a similar way overwriting S3FileLoaderIterDataPipe. By using a 'continue' instead of a 'pass' in the `MissingFileException` I achieve that the iterator completely ignores the sample and yields the next one.

Thanks for the help!

**Comment by ejguan** [comment]
Sounds great. Closing the issue as it has been solved. Free free to re-open it if you have further question.
